src/e2e/test_harness.ts

```diff
@@ -0,0 +1,485 @@
+import * as assets from "#erdstall/ledger/assets";
+import { Session } from "#erdstall";
+import * as test from "#erdstall/test";
+import {
+	PerunArt__factory,
+	PerunToken__factory,
+} from "#erdstall/ledger/backend/contracts";
+
+import { ethers, utils } from "ethers";
+import * as fs from "fs";
+import { exec } from "child_process";
+import {
+	PERUN_ADDR,
+	PART_ADDR,
+	MNEMONIC,
+	OP_PORT,
+	OP_ENDPOINT,
+	NODE_ENDPOINT,
+	NODE_PORT,
+} from "./parameters";
+import { SDKActions } from "./sdk_actions";
+
+export const ALICE = 0;
+export const BOB = 1;
+export const CHARLIE = 2;
+export const DAGOBERT = 3;
+
+export function endToEndTestHarness(sdkActions: SDKActions) {
+	return () => {
+		const rng = test.newPrng();
+		const numOfEpochs = 30;
+		const numOfAccs = 10;
+		const fundingAmountETH = 1000;
+		let erdstallProcessTerminate: Promise<void>;
+		let teeAddress: Address;
+
+		before(async function () {
+			fs.writeFileSync(
+				"operator.cfg",
+				JSON.stringify({
+					Mnemonic: MNEMONIC,
+					OperatorDerivationPath: "m/44'/60'/0'/0/0",
+					EnclaveDerivationPath: "m/44'/60'/0'/0/1",
+					EpochDuration: 3,
+					PowDepth: 0,
+					RPCPort: OP_PORT,
+					RPCHost: "0.0.0.0",
+					NodeReqTimeout: 10,
+					WaitMinedTimeout: 60,
+					NFTokenBaseURI: "http://127.0.0.1:8440/metadata/",
+				}),
+			);
+
+			return new Promise((accept, reject) => {
+				const process = exec(
+					[
+						"$ERDSTALL_EXE",
+						"-config operator.cfg",
+						"-log-level trace",
+						`-epochs ${numOfEpochs}`,
+						`-ledger-port ${NODE_PORT}`,
+						`-n ${numOfAccs}`,
+						`-funds ${fundingAmountETH}`,
+					].join(" "),
+					(e) => {
+						if (e) reject(e);
+					},
+				);
+
+				erdstallProcessTerminate = new Promise<void>(
+					(accept, reject) => {
+						process.on("close", accept);
+						process.on("error", reject);
+					},
+				);
+				const timeout = setTimeout(accept, 30000);
+				process.on("error", (e: Error) => {
+					clearTimeout(timeout);
+					reject(e);
+				});
+			});
+		});
+		after(async () => erdstallProcessTerminate);
+
+		const sessions: Session[] = [];
+		const forSessionsDo = async <R>(
+			action: (client: Session) => Promise<R>,
+			range: { from?: number; upto?: number } = {},
+		) => {
+			return Promise.all(
+				sessions.slice(range.from, range.upto).map(action),
+			);
+		};
+
+		before("create client sessions", async function () {
+			for (let i = 0; i < 4; i++) {
+				sessions[i] = await sdkActions.create(
+					NODE_ENDPOINT,
+					// We offset the accounts we retrieve from ganache by 3 so enclave,
+					// operator and potentially the bot stay untouched.
+					getGanacheSigner(i + 3),
+					OP_ENDPOINT,
+				);
+			}
+
+			sessions[ALICE].once(
+				"config",
+				(config) => (teeAddress = config.contract),
+			);
+
+			return Promise.all(sessions.map(sdkActions.initialize));
+		});
+
+		it("subscribe clients", async function () {
+			forSessionsDo(sdkActions.subscribe);
```

Comment:
Yes, we await with the `Promise` down below and subscribing is not answered with a message by the operator which is why it would return instantly anyway.